crates/turborepo-scm/src/lib.rs

```diff
@@ -21,4 +29,57 @@ pub enum Error {
     Path(#[from] PathError, #[backtrace] backtrace::Backtrace),
     #[error("could not find git binary")]
     GitBinaryNotFound(#[from] which::Error),
+    #[error("encoding error: {0}")]
+    Encoding(
+        #[from] std::string::FromUtf8Error,
+        #[backtrace] backtrace::Backtrace,
+    ),
+}
+
+impl Error {
+    pub(crate) fn git_error(s: impl Into<String>) -> Self {
+        Error::Git(s.into(), Backtrace::capture())
+    }
+
+    fn from_git_exit_code(cmd: &str, path: &AbsoluteSystemPath, exit_code: Option<i32>) -> Error {
+        let s = format!(
+            "'{}' in {} exited with status code {}",
+            cmd,
+            path.to_string(),
+            exit_code
+                .map(|code| code.to_string())
+                .unwrap_or("unknown".to_string())
+        );
+        Error::Git(s, Backtrace::capture())
+    }
+}
+
+pub(crate) fn read_git_error<R: Read>(stderr: &mut R) -> Option<Error> {
+    let mut buf = String::new();
+    if let Ok(bytes_read) = stderr.read_to_string(&mut buf) {
+        if bytes_read > 0 {
+            // something failed with git, report that error
+            Some(Error::git_error(buf))
+        } else {
+            None
+        }
+    } else {
+        None
+    }
```

Comment:
nit: this could be
```suggestion
    let bytes_read = stderr.read_to_string(&mut buf).ok()?;
    if bytes_read > 0 {
        // something failed with git, report that error
        Some(Error::git_error(buf))
    } else {
        None
    }
```

Reply:
If you *really* want to code golf you can do:

```suggestion
    let bytes_read = stderr.read_to_string(&mut buf).ok()?;
    (bytes_read > 0).then(|| Error::git_error(buf))
```

But that's a little unnecessarily fancy tbh